docs/path-resolution.rst

```diff
@@ -0,0 +1,470 @@
+
+***************
+Path Resolution
+***************
+
+In order to be able to support reproducible builds on all platforms, the Solidity compiler has to
+abstract away the details of the filesystem where source files are stored.
+Paths used in imports must work the same way everywhere while the command-line interface must be
+able to work with platform-specific paths to provide good user experience.
+This section aims to explain in detail how Solidity reconciles these requirements.
+
+.. index:: ! virtual filesystem, ! VFS, ! source unit name
+.. _virtual-filesystem:
+
+Virtual Filesystem
+==================
+
+The compiler maintains an internal database (*virtual filesystem* or *VFS* for short) where each
+source unit is assigned a unique *source unit name* which is an opaque and unstructured identifier.
+When you use the :ref:`import statement <import>`, you specify an *import path* that references a
+source unit name.
+
+.. index:: ! import callback, ! Host Filesystem Loader
+.. _import-callback:
+
+Import Callback
+---------------
+
+The VFS is initially populated only with files the compiler has received as input.
+Additional files can be loaded during compilation using an *import callback*.
+If the compiler does not find any source unit name matching the import path in the VFS, it invokes
+the callback, which is responsible for obtaining the source code to be placed under that name.
+An import callback is free to interpret source unit names in an arbitrary way, not just a paths.
+If there is no callback available when one is needed or if it fails to locate the source code,
+compilation fails.
+
+The command-line compiler provides the *Host Filesystem Loader* - a rudimentary callback
+that interprets a source unit name as a path in the local filesystem.
+The `JavaScript interface <https://github.com/ethereum/solc-js>`_ does not provide any by default,
+but one can be provided by the user.
+This mechanism can be used to obtain source code from locations other then the local filesystem
+(which may not even be accessible, e.g. when the compiler is running in a browser).
+For example the `Remix IDE <https://remix.ethereum.org/>`_ provides a versatile callback that
+lets you `import files from HTTP, IPFS and Swarm URLs or refer directly to packages in NPM registry
+<https://remix-ide.readthedocs.io/en/latest/import.html>`_.
+
+.. note::
+
+    Host Filesystem Loader's file lookup is platform-dependent.
+    For example backslashes in source unit name can be interpreted as directory separators or not
+    and the lookup can be case-sensitive or not, depending on the underlying platform.
+
+    For portability it is recommended to avoid using import paths that will work correctly only
+    with a specific import callback or only on one platform.
+
+.. index:: ! CLI path
+
+Populating the Virtual Filesystem
+---------------------------------
+
+The initial content of the VFS depends on how you invoke the compiler:
+
+#. **CLI**
```

Comment:
```suggestion
#. **solc / commandline interface**
```

Reply:
Actually, this is something that badly needs a name. What do we call the command-line interface excluding `--standard-json`? I need to refer to it pretty often and I started generally referring to it as CLI even though I know it's not precise. I just do not have a better name and I don't think "solc / commandline interface" is all that much better (`--standard-json` is also a part of solc).

"Shell interface"? "non-JSON interface"? "Command mode"?

Reply:
I think command line interface is good, I would just not use the abbreviation and also clarify how to get to it (mention "solc").

Reply:
Defining a name for us does not mean that readers will know what we mean :)

Reply:
I don't mean just define it and leave it there :) I mean also using it. And of course not just an arbitrary name completely unrelated to the thing. I think that a good name could stick. The problem is just finding one that's actually good :)